gh-107146: use f-strings for argparse internals

rodrigogiraoserrao opened the pull request:
This will close #107146.

 - This PR did not touch strings that used %-style formatting but were translated first – changing the formatting method would imply people would have to fix their translations.
 - This PR did not touch the string formatting used in user-provided strings that allow customising usage/help messages – changing them would break existing programs.

<!-- gh-issue-number: gh-107146 -->
* Issue: gh-107146
<!-- /gh-issue-number -->

rodrigogiraoserrao committed:
Use f-strings for non-public string formatting

rodrigogiraoserrao committed:
Use f-strings and add explicit length validation

The formatting done with %-style formatting was being used to check if the length of the metavar matched the action.nargs type.
This TypeError was then caught in one of the call sites of _format_args and used to raise a user-facing ValueError.
This change uses the match statement to make this validation explicit and then raises a TypeError in case the validation fails so that the remainder of the code works the same way.

rodrigogiraoserrao committed:
Add contributor

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

sunmy2019 commented on Lib/argparse.py:
```diff
@@ -521,19 +521,16 @@ def _format_action(self, action):
         elif len(action_header) <= action_width:
-            tup = self._current_indent, '', action_width, action_header
-            action_header = '%*s%-*s  ' % tup
+            action_header = f'{"":{self._current_indent}}{action_header:{action_width}}  '
```
```suggestion
            action_header = f'{"":{self._current_indent}}{action_header:<{action_width}}  '
```

sunmy2019 replied:
Did my suggestion pass the 80 char line width limit? Oh

rodrigogiraoserrao replied:
Hey @sunmy2019, excellent catch!
Because the old behaviour would align left and my f-string was aligning right, and no tests failed, does this mean I should add a test for this case?

(The line was already longer than 80 chars, it wasn't your catch that made it go over. Should I also break it apart?)

Thanks!

sunmy2019 replied:
> does this mean I should add a test for this case?

I think it might not be necessary unless it is stated in the doc

rodrigogiraoserrao replied:
I've fixed the alignment, thanks.

What are the next steps?

sunmy2019 replied:
Wait for someone with write permission. But argparse has been lack of maintainers, so no one is responsible here.

rodrigogiraoserrao replied:
Alright, cheers.

sunmy2019 reviewed (approved):
Other things LGTM

rodrigogiraoserrao committed:
Merge branch 'python:main' into argparse-fstrings

rodrigogiraoserrao committed:
Align left inside f-string.

Restores original behaviour/fixes bug.
Related review comment: https://github.com/python/cpython/pull/107152#discussion_r1271733658
